### DISCUSSION TITLE
Oauth associate unauthorized

### DISCUSSION THREAD

**Discussion by hgalytoby** [fastapi-users/fastapi-users#1154] [Q&A] [answered]
I don't know if I misunderstood the meaning or I used it wrongly, please correct me.

I follow the example
https://fastapi-users.github.io/fastapi-users/10.3/configuration/oauth/#association-router-for-authenticated-users

```py
router.include_router(
    fastapi_users.get_oauth_associate_router(microsoft_oauth_client, UserRead, SECRET),
    prefix='/auth/associate/microsoft',
    tags=['auth'],
)
```

After login use api http://localhost:8000/api/auth/associate/microsoft/authorize.

Got
`
{
  "authorization_url": "https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&client_id=<client_id>&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fapi%2Fauth%2Fassociate%2Fmicrosoft%2Fcallback&state=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI2M2Q4OTc1MzA0NDA1ZGMyNTczYzIyNmQiLCJhdWQiOiJmYXN0YXBpLXVzZXJzOm9hdXRoLXN0YXRlIiwiZXhwIjoxNjc1MTQyOTA0fQ.DLZHr4nPGrNCboRemy99hoHygUxJOl4DVyVLRMfznZg&scope=User.Read&response_mode=query"
}
`

When I do microsoft login, I end up with `{"detail":"Unauthorized"}`

I looked at the source code and found the _authenticate method.

The received kwargs are: 
`{'jwt': None, 'strategy_jwt': <fastapi_users.authentication.strategy.jwt.JWTStrategy object at 0x10babc2e0>}`

I tested to write the jwt token in it, and it is indeed feasible.
![image](https://user-images.githubusercontent.com/50397689/215666508-7a41a6aa-3709-4121-948d-5194b19d85c7.png)

I don't know what I did wrong, the jwt token was not received, I also tried to use github oauth, the result is the same.

I know the associate callback has `Depends(get_current_active_user)`, but the oauth callback doesn't give me the bearer token.

How should I make the callback put the bearer token back?

Could you please help me?

**Comment by frankie567** [comment, marked as answer]
The associate router is used to link existing users to an OAuth provider. For this to work, the user needs to be *authenticated* when calling `/auth/associate/microsoft/callback`.

Since you're using a JWT token, it can't work transparently with the redirection (you would have to use a Cookie instead). I suppose you have a frontend (like a JS or mobile app) which stores the JWT token somewhere.

The solution then is to have an intermediate page *on your frontend* where the user will be redirected after Microsoft login (instead of `/auth/associate/microsoft/callback`). At this point, you'll be able to catch the `code` in the query parameters and manually make an API request to `/auth/associate/microsoft/callback`, taking care of setting the JWT token you have in memory.

**Reply by hgalytoby** [reply]
Thank you I understand.
